Samples CMake update part 2

aserdean opened the pull request:
- Add CMake build for file-encryptor
- Add CMake build for helloworld
- Add CMake build for local attestation
- Add CMake build for remote attestation
- Add CMake file-encryptor helloworld local_attestation remote_attestation to test build

andschwa commented on samples/test-samples.cmake:
```diff
@@ -20,7 +20,7 @@ execute_process(COMMAND ${CMAKE_COMMAND} -E env DESTDIR=${BUILD_DIR}/install ${C
 set(INSTALL_DIR ${BUILD_DIR}/install${PREFIX_DIR})
 
 # TODO: Add the rest of the samples.
```
Remove this TODO if we've added the rest of the samples 😉

aserdean replied:
:)

andschwa commented on samples/test-samples.cmake:
```diff
@@ -38,9 +38,18 @@ foreach (SAMPLE data-sealing)
-  execute_process(
-    COMMAND ${CMAKE_COMMAND} --build ${SAMPLE_BUILD_DIR} --target run
-    RESULT_VARIABLE TEST_RESULT)
+  if (NOT DEFINED $ENV{OE_SIMULATION})
```
Is this true if `OE_SIMULATION=0`? I think I've seen that variable used as a 1/0 flag.

Also, it would probably be a bit cleaner to simulation semantics via `-D` to this script instead of as an environment variable.

aserdean replied:
Updated.

andschwa commented on samples/test-samples.cmake:
```diff
@@ -38,9 +38,18 @@ foreach (SAMPLE data-sealing)
+  endif ()
+
+  if (${SAMPLE} MATCHES "(file-encryptor|helloworld)")
+    # Simulation tests
```
This comment doesn't add much value. Maybe, "Of all the current samples, only `file-encryptor` and `helloworld` also work under simulation, so we test that additional scenario here."

Otherwise it's not that clear why we're doing this.

aserdean replied:
Sorry about it, I folded in you comments and added you as co-author.

andschwa commented on samples/test-samples.cmake:
```diff
@@ -38,9 +38,18 @@ foreach (SAMPLE data-sealing)
+    # Simulation tests
+    execute_process(
+      COMMAND ${CMAKE_COMMAND} --build ${SAMPLE_BUILD_DIR} --target simulate_run
+      RESULT_VARIABLE TEST_RESULT)
```
This will overwrite the result of the variable above, leading to missed failures if the non-simulation run fails, but the simulation runs.

aserdean replied:
Oops, thanks for catching this one, I missed it.

aserdean commented on samples/file-encryptor/host/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+add_executable(file-encryptor_host host.cpp ${CMAKE_CURRENT_BINARY_DIR}/fileencryptor_u.c)
+
+target_include_directories(file-encryptor_host PRIVATE
+  ${CMAKE_CURRENT_SOURCE_DIR} # For shared.h
```
The include looks like the following:
```
/home/azureuser/test/openenclave/build/samples/file-encryptor/host/fileencryptor_args.h:12:10: fatal error: ../shared.h: No such file or directory
2:  #include "../shared.h"
```

I will update the comment

aserdean commented on samples/helloworld/enc/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+
+target_include_directories(enclave PRIVATE ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
+
+target_link_libraries(enclave openenclave::oeenclave openenclave::oelibcxx)
```
Sorry about it, you are right it only needs oelibc

aserdean commented on samples/file-encryptor/enc/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+
+target_compile_definitions(enclave PUBLIC OE_API_VERSION=2)
+
+target_include_directories(enclave PRIVATE ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
```
Again we need this one for:
```
/home/azureuser/test/openenclave/build/samples/file-encryptor/enc/fileencryptor_args.h:12:10: fatal error: ../shared.h: No such file or directory
2:  #include "../shared.h"
```
I'll double check the rest as well.

andschwa replied:
Oof, oh well I guess.

aserdean commented on samples/local_attestation/common/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+add_custom_command(OUTPUT localattestation_t.h localattestation_t.c
+  COMMAND openenclave::oeedger8r --trusted ${CMAKE_SOURCE_DIR}/localattestation.edl)
+
+# Create a library common to each of our three enclaves.
```
Sloppiness on my side :(.

I'll update the comment on the remote attestation as well.

aserdean commented on samples/file-encryptor/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+    COMMAND file-encryptor_host ${CMAKE_CURRENT_SOURCE_DIR}/testfile ${CMAKE_BINARY_DIR}/enc/enclave.signed)
+endif ()
+
+add_custom_target(simulate_run
```
I thought `simulate_run` would be more appropriate for some reason, but indeed `simulate` sounds better.

I will update it.

aserdean commented on samples/local_attestation/CMakeLists.txt:
```diff
@@ -0,0 +1,53 @@
+# Generate header for A
+add_custom_command(OUTPUT enc2/enc1_pubkey.h
+  DEPENDS ${CMAKE_BINARY_DIR}/public_a.pem
+  COMMAND chmod u+x ${CMAKE_SOURCE_DIR}/gen_pubkey_header.sh
```
The execute permission in the repo is good, however the one distributed via the install target isn't.
i.e:
```
execvp: /home/azureuser/test/openenclave/build/install/opt/openenclave/share/openenclave/samples/local_attestation/gen_pubkey_header.sh: Permission denied
2: CMakeFiles/local_attestation_headers.dir/build.make:61: recipe for target 'enc2/enc1_pubkey.h' failed
```
the same applies for the Makefile:
https://github.com/Microsoft/openenclave/blob/master/samples/local_attestation/enc1/Makefile#L44

I wanted to fold in the fold in the following diff but was unsure:
```
git diff samples/CMakeLists.txt
diff --git a/samples/CMakeLists.txt b/samples/CMakeLists.txt
index 1345a21d9..3981010e1 100644
--- a/samples/CMakeLists.txt
+++ b/samples/CMakeLists.txt
@@ -3,7 +3,8 @@

 if (USE_LIBSGX)
   install(DIRECTORY remote_attestation local_attestation
-    DESTINATION ${CMAKE_INSTALL_DATADIR}/openenclave/samples)
+    DESTINATION ${CMAKE_INSTALL_DATADIR}/openenclave/samples
+    USE_SOURCE_PERMISSIONS)
 endif ()

 install(DIRECTORY helloworld file-encryptor data-sealing
```

Please let me know if I should include above incremental.

andschwa replied:
I think `USE_SOURCE_PERMISSIONS` is a better way to go about this than scripting in chmod 😄 be happy to take that.

aserdean replied:
I updated the PR with both CMake and Makefiles changes.

aserdean commented on samples/local_attestation/enc1/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+add_executable(enclave_a ecalls.cpp)
+target_link_libraries(enclave_a common)
+target_include_directories(enclave_a PRIVATE ${CMAKE_BINARY_DIR}/enc1)
```
I messed things a bit here because I was confused by:
https://github.com/Microsoft/openenclave/blob/master/samples/local_attestation/Makefile#L6-L8
 and thought it was a good idea to make it top level.

I will update the to be per enclave basis.

Also should I send out a PR to update the Makefiles as well? The top level genkey target is not used.

andschwa replied:
If it can be per enclave, it makes some things easier; but sometimes they can't be.

If you find stuff that can be fixed in the Makefiles, go for it! PRs are welcome.

aserdean replied:
I will send out a PR to fan out those targets from the Makefiles.

I updated the this PR with the key modifications per enclave can you PTAL?

andschwa commented on samples/local_attestation/CMakeLists.txt:
```diff
@@ -0,0 +1,53 @@
+  COMMAND openssl rsa -in private_b.pem -pubout -out public_b.pem)
+
+# Generate header for A
+add_custom_command(OUTPUT enc2/enc1_pubkey.h
```
Resolving as there's a long email thread about this.

aserdean commented on samples/file-encryptor/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+  DEPENDS ${CMAKE_BINARY_DIR}/private.pem
+  COMMAND openenclave::oesign sign $<TARGET_FILE:enclave> ${CMAKE_SOURCE_DIR}/enc/file-encryptor.conf private.pem)
+
+if (NOT DEFINED $ENV{OE_SIMULATION})
```
I wanted to let the build and simulation tests to run if OE_SIMULATION is specified. I didn't thought that one would define it to zero but I updated the PR to account for that.

andschwa commented on samples/file-encryptor/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+  DEPENDS ${CMAKE_BINARY_DIR}/private.pem
+  COMMAND openenclave::oesign sign $<TARGET_FILE:enclave> ${CMAKE_SOURCE_DIR}/enc/file-encryptor.conf private.pem)
+
+if ((NOT DEFINED ENV{OE_SIMULATION}) OR (NOT $ENV{OE_SIMULATION}))
```
Wait, this confuses me more. Can we just take `OE_SIMULATION` as `-DOE_SIMULATION=ON` (or `OFF`) when we call the script, and then check `if (OE_SIMULATION)` here; I really don't like adding an extra vector to send information to this script.

aserdean replied:
That would be ideal, but we have to keep in mind that the [Docs](https://github.com/Microsoft/openenclave/blob/e0d3ce41926287d418ba44f403c651c045fcc76b/docs/GettingStartedDocs/SimulatorGettingStarted.md#run-unittests) and [code](https://github.com/Microsoft/openenclave/blob/710702db7ca1a0bfdd45deb53c74f5c7a784deff/host/tests.c#L15) are using it as an environment variable. Creating a special case with `-DOE_SIMULATION=ON/OFF` just for the samples would add more confusion in my opinion.
We can change the usage to a `-D` as whole but not part of this effort.

andschwa replied:
Fair enough!

andschwa replied:
Filed #1587.

andschwa commented on samples/helloworld/host/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+
+add_executable(helloworld_host host.c ${CMAKE_CURRENT_BINARY_DIR}/helloworld_u.c)
+
+target_include_directories(helloworld_host PRIVATE
```
Since it is a sample, let's keep being super explicit:

`# Need for the generated file helloworld_u.h`

aserdean replied:
Updated.

andschwa commented on samples/local_attestation/enc1/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+  COMMAND ${CMAKE_SOURCE_DIR}/gen_pubkey_header.sh ${CMAKE_BINARY_DIR}/enc2/enc1_pubkey.h ${CMAKE_BINARY_DIR}/public_a.pem)
+
+add_custom_target(enc1_pubkey_h DEPENDS enc1_pubkey.h)
+add_custom_target(key_info_a DEPENDS public_key_a enc2_pubkey_h)
```
Hm... I think I can simplify this but it's too hard to do in a code review. I'll give it a shot and share a commit in a bit.

andschwa commented on samples/test-samples.cmake:
```diff
@@ -38,16 +35,26 @@ foreach (SAMPLE data-sealing)
   # failed sample will cause the rest of the samples to be skipped.
-  if (TEST_RESULT)
+  if (TEST_RESULT OR TEST_SIMULATE_RESULT)
     message(FATAL_ERROR "Samples test '${SAMPLE}' failed!")
```
But did it fail in simulation or not?

andschwa replied:
I think a simpler approach would just do an `if` check after each run.

andschwa commented:
I don't think I can push to your branch (by the way, feel free to develop in this repo, it's a bit easier to collaborate on branches that way), but here's a cleanup of the targets and key generation for `local_attestation`: [d52c2e5841481bab8de6c022790595c3c6aaa0e9](https://github.com/Microsoft/openenclave/commit/d52c2e5841481bab8de6c022790595c3c6aaa0e9)

Please also feel free to cherry-pick it (and @CodeMonkeyLeet you may want to take a look, I think it's close to what you were after in your email thread, though I kept the key generation next to the enclaves that needed those keys just to keep the sample simple).

aserdean commented:
> I don't think I can push to your branch (by the way, feel free to develop in this repo, it's a bit easier to collaborate on branches that way), but here's a cleanup of the targets and key generation for `local_attestation`: [d52c2e5841481bab8de6c022790595c3c6aaa0e9](https://github.com/Microsoft/openenclave/commit/d52c2e5841481bab8de6c022790595c3c6aaa0e9)
Call me old fashioned but I prefer using diffs or ML format (`git format-patch`) and applying it with `git am` :)
> 
> Please also feel free to cherry-pick it (and @CodeMonkeyLeet you may want to take a look, I think it's close to what you were after in your email thread, though I kept the key generation next to the enclaves that needed those keys just to keep the sample simple).
I cherry-picked your commit and did the same for remote attestation.

Originally I did the same but I was unsure if I changed the key generation for local/remote attestation inside the enclaves should that be applied to the rest of the samples 😄?

andschwa commented:
> Call me old fashioned but I prefer using diffs or ML format (`git format-patch`) and applying it with `git am`, I usually avoid creating branches in the source repo because I forget to delete the branches and I end up polluting it 😆 .

That's totally fine.

> Originally I did the same but I was unsure if I changed the key generation for local/remote attestation inside the enclaves should that be applied to the rest of the samples 😄?

I am down for whatever is cleanest. The goal is for the samples to be as canonical and straightforward as possible.

andschwa commented on samples/file-encryptor/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+cmake_minimum_required(VERSION 3.10)
```
So sorry... since you've found the real minimum required version, can you bump them all to 3.11? I'm fine if you want to bump that to another PR too.

CodeMonkeyLeet replied:
@anakrish IIRC we chose some version of cmake that's not the latest for compatibility with the cmake version that Visual Studio deploys on Windows. Is the bump to 3.11 consistent with that?

anakrish replied:
3.12 is the version that we have locked down for OE due to Visual Studio.
https://github.com/Microsoft/openenclave/commit/8121f4606c73982566276cee89e4df256342319c.

Any reason why we want to choose a lesser cmake version for samples?

andschwa replied:
My reasoning would be to make it as easy as possible to use the samples.

aserdean replied:
Since Visual Studio comes with `3.12` (as per https://github.com/Microsoft/openenclave/commit/8121f4606c73982566276cee89e4df256342319c), I have bumped the version to `3.11` for the samples because that is the minimum we need for Linux (https://github.com/Microsoft/openenclave/issues/1561)

We can bump it to 3.12 if needed with a different PR.

andschwa commented on samples/helloworld/enc/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+# Need for the generated file helloworld_t.h
+target_include_directories(enclave PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+
+target_link_libraries(enclave openenclave::oeenclave openenclave::oelibc)
```
Ah I just updated the documentation for this in #1585 😄 Sorry about that, dependencies got changed underneath me!

andschwa commented on samples/local_attestation/enc1/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
+target_include_directories(enclave_a PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+
+target_link_libraries(enclave_a common)
```
Ah, I fucked up the naming for these in data-sealing. Filed #1588.

andschwa commented on samples/test-samples.cmake:
```diff
@@ -38,18 +35,31 @@ foreach (SAMPLE data-sealing)
+    endif ()
+  endif()
 
   # The prior common cannot succeed unless all commands before also
```
What did I mean here? The prior common? I think I mean "The prior code." 😅

aserdean replied:
😄 Updated.

andschwa commented on samples/test-samples.cmake:
```diff
@@ -38,18 +35,31 @@ foreach (SAMPLE data-sealing)
-  if (TEST_RESULT)
-    message(FATAL_ERROR "Samples test '${SAMPLE}' failed!")
-  endif ()
+  # TODO: An unfortunate side-effect of using FATAL_ERROR and its placement
```
Filed #1589.

aserdean replied:
I added a variable to store the result if one sample failed.

Not the finest solution but it should work.

andschwa commented on samples/test-samples.cmake:
```diff
@@ -38,18 +35,31 @@ foreach (SAMPLE data-sealing)
+  # TODO: An unfortunate side-effect of using FATAL_ERROR and its placement
+  # is that a failed sample will cause the rest of the samples to be skipped.
 
   # TODO: Build the sample with GNU Make.
```
Filed #1590.

andschwa commented:
@aserdean I just filed #1594 after @paulcallen got that rebuilding wasn't working after he edited the EDL file. I forgot the `DEPENDS` line. Think it should go into this PR?

CodeMonkeyLeet commented on samples/local_attestation/common/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+add_library(common STATIC attestation.cpp crypto.cpp dispatcher.cpp ${CMAKE_CURRENT_BINARY_DIR}/localattestation_t.c)
+target_compile_definitions(common PUBLIC OE_API_VERSION=2)
+target_link_libraries(common PUBLIC
+  # `liboecore`, a dependency of `liboeenclave`, requires the ecalls
```
@andschwa I realize this is carried over from the already merged data_sealing sample, but we might want to clean this comment up as part of the doc-a-thon; it reads like it was intended for a maintainer audience (such as the vague suggestion for the alternative.)

andschwa replied:
Yeah, probably!

aserdean commented:
> @aserdean I just filed #1594 after @paulcallen got that rebuilding wasn't working after he edited the EDL file. I forgot the `DEPENDS` line. Think it should go into this PR?

I updated the PR to include it.

aserdean commented:
@andschwa , @johnkord I updated the PR and addressed a few more issues. If everything looks fine can we get this merged?

I want to address: https://github.com/Microsoft/openenclave/issues/1588 and https://github.com/Microsoft/openenclave/issues/1590 in a different PR, as the current effort addresses more things that originally planned.

aserdean commented:
bors try

andschwa commented:
bors r+
